File: static/js/event_bus.js
/**
 * Frontend event bus — mirrors the Python-side pattern.
 * Components communicate through events rather than direct coupling.
 *
 * Usage:
 *   EventBus.on('contacts:changed', handler)
 *   EventBus.emit('contacts:changed', { count: 5 })
 *   EventBus.off('contacts:changed', handler)
 */
const EventBus = (() => {
  const _listeners = {};
  const _wildcards = [];

  function on(event, callback) {
    if (!_listeners[event]) _listeners[event] = [];
    _listeners[event].push(callback);
  }

  function off(event, callback) {
    if (!_listeners[event]) return;
    _listeners[event] = _listeners[event].filter(cb => cb !== callback);
  }

  // Register a listener for ALL events. Returns an unsubscribe function.
  function onAny(callback) {
    _wildcards.push(callback);
    return () => {
      const i = _wildcards.indexOf(callback);
      if (i >= 0) _wildcards.splice(i, 1);
    };
  }

  function emit(event, data) {
    _wildcards.forEach(fn => { try { fn(event, data); } catch {} });
    (_listeners[event] || []).forEach(cb => {
      try { cb(data); }
      catch (e) { console.error(`EventBus listener error on '${event}':`, e); }
    });
  }

  function clear(event) {
    if (event) delete _listeners[event];
    else Object.keys(_listeners).forEach(k => delete _listeners[k]);
  }

  return { on, off, onAny, emit, clear };
})();

File: core/event_bus.py
"""Simple synchronous pub/sub event bus (Python side).

Usage:
    from core.event_bus import bus

    bus.on("contacts:updated", lambda data: print(data))
    bus.emit("contacts:updated", {"count": 5})
    bus.off("contacts:updated", my_handler)
"""
from collections import defaultdict
from typing import Any, Callable

from core.logger import get_logger

logger = get_logger("event_bus")


class EventBus:
    def __init__(self) -> None:
        self._listeners: dict[str, list[Callable]] = defaultdict(list)

    def on(self, event: str, callback: Callable) -> None:
        self._listeners[event].append(callback)
        logger.debug(f"Subscribed '{callback.__name__}' to '{event}'")

    def off(self, event: str, callback: Callable) -> None:
        self._listeners[event] = [
            cb for cb in self._listeners[event] if cb is not callback
        ]

    def emit(self, event: str, data: Any = None) -> None:
        logger.debug(f"Emit '{event}' data={data!r}")
        for cb in list(self._listeners.get(event, [])):
            try:
                cb(data)
            except Exception as exc:
                logger.error(f"Listener '{cb.__name__}' raised on '{event}': {exc}")

    def clear(self, event: str | None = None) -> None:
        if event:
            self._listeners.pop(event, None)
        else:
            self._listeners.clear()


# Application-wide singleton
bus = EventBus()
